In [115]:
import pandas as pd

In [116]:
# Read in the first CSV:
df1 = pd.read_csv("fake_data_address.csv")
df1.head(5)

/var/folders/jd/f52yh_4n4md3vqk92c4p3s3w0000gn/T/ipykernel_57298/306956661.py:2: DtypeWarning: Columns (1,2,3) have mixed types. Specify dtype option on import or set low_memory=False.
  df1 = pd.read_csv("fake_data_address.csv")


id first_name middle_name last_name  \
0  1.0   Jerrylee      Terrel   Sanchis   
1  2.0   Florenza      Jethro   Shapera   
2  3.0    Agustin        Juan      Mart   
3  4.0      Pammi      Renell     Ticic   
4  5.0       Conn      Tybalt   Sprasen   

               Line 1: address Line 2: address2   state-city      zip  
0          Line 1: 4 Continental Trail Line 2:     IL-Peoria  61614.0  
1    Line 1: 0259 Messerschmidt Street Line 2:       PA-Erie  16565.0  
2  Line 1: 3461 Bartillon Lane Line 2: Suite 26  CA-Van Nuys  91499.0  
3  Line 1: 78266 Lien Road Line 2: PO Box 68671   PA-Reading  19610.0  
4      Line 1: 4392 Duke Place Line 2: Suite 79   WA-Spokane  99215.0

In [117]:
df2 = pd.read_csv("fake_data_dob.csv")
df2.head(5)

/var/folders/jd/f52yh_4n4md3vqk92c4p3s3w0000gn/T/ipykernel_57298/871118560.py:1: DtypeWarning: Columns (1,2,3) have mixed types. Specify dtype option on import or set low_memory=False.
  df2 = pd.read_csv("fake_data_dob.csv")


last_name, first_name middle_name         dob                        email  \
0          Sanchis, Jerrylee Terrel  11/20/1924          tsanchis0@google.cn   
1          Shapera, Florenza Jethro   3/14/1923  jshapera1@howstuffworks.com   
2                Mart, Agustin Juan  12/19/1941           jmart2@yahoo.co.jp   
3               Ticic, Pammi Renell    8/2/1979              rticic3@php.net   
4              Sprasen, Conn Tybalt   4/14/1966    tsprasen4@dagondesign.com   

                                   guid  
0  38b23627-9118-4180-b0d5-1396b8c99056  
1  70f362c6-ad87-4e31-aa51-1f829fcdcebb  
2  6b4cf83d-a760-4546-9b2a-2bbe146007f3  
3  31c21b5d-7cdd-4acd-8592-b156b2e519c7  
4  ecd3fec5-e4ec-45ff-91b0-bb0db86596a9

In [118]:
# Create a 'full_name' column that we can use to merge the two dataframes:
df1 = df1.assign(full_name = df1.last_name + ", " + df1.first_name + " " + df1.middle_name) 
df1.head(5)

id first_name middle_name last_name  \
0  1.0   Jerrylee      Terrel   Sanchis   
1  2.0   Florenza      Jethro   Shapera   
2  3.0    Agustin        Juan      Mart   
3  4.0      Pammi      Renell     Ticic   
4  5.0       Conn      Tybalt   Sprasen   

               Line 1: address Line 2: address2   state-city      zip  \
0          Line 1: 4 Continental Trail Line 2:     IL-Peoria  61614.0   
1    Line 1: 0259 Messerschmidt Street Line 2:       PA-Erie  16565.0   
2  Line 1: 3461 Bartillon Lane Line 2: Suite 26  CA-Van Nuys  91499.0   
3  Line 1: 78266 Lien Road Line 2: PO Box 68671   PA-Reading  19610.0   
4      Line 1: 4392 Duke Place Line 2: Suite 79   WA-Spokane  99215.0   

                  full_name  
0  Sanchis, Jerrylee Terrel  
1  Shapera, Florenza Jethro  
2        Mart, Agustin Juan  
3       Ticic, Pammi Renell  
4      Sprasen, Conn Tybalt

In [119]:
# Rename unweildy column name to 'full_name':
df2['full_name'] = df2['last_name, first_name middle_name']
df2.head(5)

last_name, first_name middle_name         dob                        email  \
0          Sanchis, Jerrylee Terrel  11/20/1924          tsanchis0@google.cn   
1          Shapera, Florenza Jethro   3/14/1923  jshapera1@howstuffworks.com   
2                Mart, Agustin Juan  12/19/1941           jmart2@yahoo.co.jp   
3               Ticic, Pammi Renell    8/2/1979              rticic3@php.net   
4              Sprasen, Conn Tybalt   4/14/1966    tsprasen4@dagondesign.com   

                                   guid                 full_name  
0  38b23627-9118-4180-b0d5-1396b8c99056  Sanchis, Jerrylee Terrel  
1  70f362c6-ad87-4e31-aa51-1f829fcdcebb  Shapera, Florenza Jethro  
2  6b4cf83d-a760-4546-9b2a-2bbe146007f3        Mart, Agustin Juan  
3  31c21b5d-7cdd-4acd-8592-b156b2e519c7       Ticic, Pammi Renell  
4  ecd3fec5-e4ec-45ff-91b0-bb0db86596a9      Sprasen, Conn Tybalt

In [120]:
# Merge the two dataframes on the 'full_name' column:
df3 = pd.merge(df1, df2, on='full_name')
df3.head(5)

id first_name middle_name last_name  \
0  1.0   Jerrylee      Terrel   Sanchis   
1  2.0   Florenza      Jethro   Shapera   
2  3.0    Agustin        Juan      Mart   
3  4.0      Pammi      Renell     Ticic   
4  5.0       Conn      Tybalt   Sprasen   

               Line 1: address Line 2: address2   state-city      zip  \
0          Line 1: 4 Continental Trail Line 2:     IL-Peoria  61614.0   
1    Line 1: 0259 Messerschmidt Street Line 2:       PA-Erie  16565.0   
2  Line 1: 3461 Bartillon Lane Line 2: Suite 26  CA-Van Nuys  91499.0   
3  Line 1: 78266 Lien Road Line 2: PO Box 68671   PA-Reading  19610.0   
4      Line 1: 4392 Duke Place Line 2: Suite 79   WA-Spokane  99215.0   

                  full_name last_name, first_name middle_name         dob  \
0  Sanchis, Jerrylee Terrel          Sanchis, Jerrylee Terrel  11/20/1924   
1  Shapera, Florenza Jethro          Shapera, Florenza Jethro   3/14/1923   
2        Mart, Agustin Juan                Mart, Agustin Juan  12/19/1941   
3       Ticic, Pammi Renell               Ticic, Pammi Renell    8/2/1979   
4      Sprasen, Conn Tybalt              Sprasen, Conn Tybalt   4/14/1966   

                         email                                  guid  
0          tsanchis0@google.cn  38b23627-9118-4180-b0d5-1396b8c99056  
1  jshapera1@howstuffworks.com  70f362c6-ad87-4e31-aa51-1f829fcdcebb  
2           jmart2@yahoo.co.jp  6b4cf83d-a760-4546-9b2a-2bbe146007f3  
3              rticic3@php.net  31c21b5d-7cdd-4acd-8592-b156b2e519c7  
4    tsprasen4@dagondesign.com  ecd3fec5-e4ec-45ff-91b0-bb0db86596a9

In [121]:
# Create a new 'line1' address column for line 1 of the address:
re =  r'^Line 1: (.*)(?:Line 2: )'
df3['line1'] = df3['Line 1: address Line 2: address2'].str.extract(re)
df3.head(5)

id first_name middle_name last_name  \
0  1.0   Jerrylee      Terrel   Sanchis   
1  2.0   Florenza      Jethro   Shapera   
2  3.0    Agustin        Juan      Mart   
3  4.0      Pammi      Renell     Ticic   
4  5.0       Conn      Tybalt   Sprasen   

               Line 1: address Line 2: address2   state-city      zip  \
0          Line 1: 4 Continental Trail Line 2:     IL-Peoria  61614.0   
1    Line 1: 0259 Messerschmidt Street Line 2:       PA-Erie  16565.0   
2  Line 1: 3461 Bartillon Lane Line 2: Suite 26  CA-Van Nuys  91499.0   
3  Line 1: 78266 Lien Road Line 2: PO Box 68671   PA-Reading  19610.0   
4      Line 1: 4392 Duke Place Line 2: Suite 79   WA-Spokane  99215.0   

                  full_name last_name, first_name middle_name         dob  \
0  Sanchis, Jerrylee Terrel          Sanchis, Jerrylee Terrel  11/20/1924   
1  Shapera, Florenza Jethro          Shapera, Florenza Jethro   3/14/1923   
2        Mart, Agustin Juan                Mart, Agustin Juan  12/19/1941   
3       Ticic, Pammi Renell               Ticic, Pammi Renell    8/2/1979   
4      Sprasen, Conn Tybalt              Sprasen, Conn Tybalt   4/14/1966   

                         email                                  guid  \
0          tsanchis0@google.cn  38b23627-9118-4180-b0d5-1396b8c99056   
1  jshapera1@howstuffworks.com  70f362c6-ad87-4e31-aa51-1f829fcdcebb   
2           jmart2@yahoo.co.jp  6b4cf83d-a760-4546-9b2a-2bbe146007f3   
3              rticic3@php.net  31c21b5d-7cdd-4acd-8592-b156b2e519c7   
4    tsprasen4@dagondesign.com  ecd3fec5-e4ec-45ff-91b0-bb0db86596a9   

                        line1  
0        4 Continental Trail   
1  0259 Messerschmidt Street   
2        3461 Bartillon Lane   
3            78266 Lien Road   
4            4392 Duke Place

In [122]:
# Create a new 'line2' column for the second line of the address:
re =  r'.*Line 2: (.*)'
df3['line2'] = df3['Line 1: address Line 2: address2'].str.extract(re)
df3.head(5)

id first_name middle_name last_name  \
0  1.0   Jerrylee      Terrel   Sanchis   
1  2.0   Florenza      Jethro   Shapera   
2  3.0    Agustin        Juan      Mart   
3  4.0      Pammi      Renell     Ticic   
4  5.0       Conn      Tybalt   Sprasen   

               Line 1: address Line 2: address2   state-city      zip  \
0          Line 1: 4 Continental Trail Line 2:     IL-Peoria  61614.0   
1    Line 1: 0259 Messerschmidt Street Line 2:       PA-Erie  16565.0   
2  Line 1: 3461 Bartillon Lane Line 2: Suite 26  CA-Van Nuys  91499.0   
3  Line 1: 78266 Lien Road Line 2: PO Box 68671   PA-Reading  19610.0   
4      Line 1: 4392 Duke Place Line 2: Suite 79   WA-Spokane  99215.0   

                  full_name last_name, first_name middle_name         dob  \
0  Sanchis, Jerrylee Terrel          Sanchis, Jerrylee Terrel  11/20/1924   
1  Shapera, Florenza Jethro          Shapera, Florenza Jethro   3/14/1923   
2        Mart, Agustin Juan                Mart, Agustin Juan  12/19/1941   
3       Ticic, Pammi Renell               Ticic, Pammi Renell    8/2/1979   
4      Sprasen, Conn Tybalt              Sprasen, Conn Tybalt   4/14/1966   

                         email                                  guid  \
0          tsanchis0@google.cn  38b23627-9118-4180-b0d5-1396b8c99056   
1  jshapera1@howstuffworks.com  70f362c6-ad87-4e31-aa51-1f829fcdcebb   
2           jmart2@yahoo.co.jp  6b4cf83d-a760-4546-9b2a-2bbe146007f3   
3              rticic3@php.net  31c21b5d-7cdd-4acd-8592-b156b2e519c7   
4    tsprasen4@dagondesign.com  ecd3fec5-e4ec-45ff-91b0-bb0db86596a9   

                        line1         line2  
0        4 Continental Trail                 
1  0259 Messerschmidt Street                 
2        3461 Bartillon Lane       Suite 26  
3            78266 Lien Road   PO Box 68671  
4            4392 Duke Place       Suite 79

In [123]:
# Create a new 'city' column for the city:
re = r'-(.*)'
df3['city'] = df3['state-city'].str.extract(re)
df3.head(5)

id first_name middle_name last_name  \
0  1.0   Jerrylee      Terrel   Sanchis   
1  2.0   Florenza      Jethro   Shapera   
2  3.0    Agustin        Juan      Mart   
3  4.0      Pammi      Renell     Ticic   
4  5.0       Conn      Tybalt   Sprasen   

               Line 1: address Line 2: address2   state-city      zip  \
0          Line 1: 4 Continental Trail Line 2:     IL-Peoria  61614.0   
1    Line 1: 0259 Messerschmidt Street Line 2:       PA-Erie  16565.0   
2  Line 1: 3461 Bartillon Lane Line 2: Suite 26  CA-Van Nuys  91499.0   
3  Line 1: 78266 Lien Road Line 2: PO Box 68671   PA-Reading  19610.0   
4      Line 1: 4392 Duke Place Line 2: Suite 79   WA-Spokane  99215.0   

                  full_name last_name, first_name middle_name         dob  \
0  Sanchis, Jerrylee Terrel          Sanchis, Jerrylee Terrel  11/20/1924   
1  Shapera, Florenza Jethro          Shapera, Florenza Jethro   3/14/1923   
2        Mart, Agustin Juan                Mart, Agustin Juan  12/19/1941   
3       Ticic, Pammi Renell               Ticic, Pammi Renell    8/2/1979   
4      Sprasen, Conn Tybalt              Sprasen, Conn Tybalt   4/14/1966   

                         email                                  guid  \
0          tsanchis0@google.cn  38b23627-9118-4180-b0d5-1396b8c99056   
1  jshapera1@howstuffworks.com  70f362c6-ad87-4e31-aa51-1f829fcdcebb   
2           jmart2@yahoo.co.jp  6b4cf83d-a760-4546-9b2a-2bbe146007f3   
3              rticic3@php.net  31c21b5d-7cdd-4acd-8592-b156b2e519c7   
4    tsprasen4@dagondesign.com  ecd3fec5-e4ec-45ff-91b0-bb0db86596a9   

                        line1         line2      city  
0        4 Continental Trail                   Peoria  
1  0259 Messerschmidt Street                     Erie  
2        3461 Bartillon Lane       Suite 26  Van Nuys  
3            78266 Lien Road   PO Box 68671   Reading  
4            4392 Duke Place       Suite 79   Spokane

In [124]:
# Create a new 'state' column for the state:
re = r'(.*)-'
df3['state'] = df3['state-city'].str.extract(re)
df3.head(5)

id first_name middle_name last_name  \
0  1.0   Jerrylee      Terrel   Sanchis   
1  2.0   Florenza      Jethro   Shapera   
2  3.0    Agustin        Juan      Mart   
3  4.0      Pammi      Renell     Ticic   
4  5.0       Conn      Tybalt   Sprasen   

               Line 1: address Line 2: address2   state-city      zip  \
0          Line 1: 4 Continental Trail Line 2:     IL-Peoria  61614.0   
1    Line 1: 0259 Messerschmidt Street Line 2:       PA-Erie  16565.0   
2  Line 1: 3461 Bartillon Lane Line 2: Suite 26  CA-Van Nuys  91499.0   
3  Line 1: 78266 Lien Road Line 2: PO Box 68671   PA-Reading  19610.0   
4      Line 1: 4392 Duke Place Line 2: Suite 79   WA-Spokane  99215.0   

                  full_name last_name, first_name middle_name         dob  \
0  Sanchis, Jerrylee Terrel          Sanchis, Jerrylee Terrel  11/20/1924   
1  Shapera, Florenza Jethro          Shapera, Florenza Jethro   3/14/1923   
2        Mart, Agustin Juan                Mart, Agustin Juan  12/19/1941   
3       Ticic, Pammi Renell               Ticic, Pammi Renell    8/2/1979   
4      Sprasen, Conn Tybalt              Sprasen, Conn Tybalt   4/14/1966   

                         email                                  guid  \
0          tsanchis0@google.cn  38b23627-9118-4180-b0d5-1396b8c99056   
1  jshapera1@howstuffworks.com  70f362c6-ad87-4e31-aa51-1f829fcdcebb   
2           jmart2@yahoo.co.jp  6b4cf83d-a760-4546-9b2a-2bbe146007f3   
3              rticic3@php.net  31c21b5d-7cdd-4acd-8592-b156b2e519c7   
4    tsprasen4@dagondesign.com  ecd3fec5-e4ec-45ff-91b0-bb0db86596a9   

                        line1         line2      city state  
0        4 Continental Trail                   Peoria    IL  
1  0259 Messerschmidt Street                     Erie    PA  
2        3461 Bartillon Lane       Suite 26  Van Nuys    CA  
3            78266 Lien Road   PO Box 68671   Reading    PA  
4            4392 Duke Place       Suite 79   Spokane    WA

In [125]:
# Convert the zip to an int, so it prints better:
df3['zip'] = df3['zip'].astype(int)
df3.head(5)

id first_name middle_name last_name  \
0  1.0   Jerrylee      Terrel   Sanchis   
1  2.0   Florenza      Jethro   Shapera   
2  3.0    Agustin        Juan      Mart   
3  4.0      Pammi      Renell     Ticic   
4  5.0       Conn      Tybalt   Sprasen   

               Line 1: address Line 2: address2   state-city    zip  \
0          Line 1: 4 Continental Trail Line 2:     IL-Peoria  61614   
1    Line 1: 0259 Messerschmidt Street Line 2:       PA-Erie  16565   
2  Line 1: 3461 Bartillon Lane Line 2: Suite 26  CA-Van Nuys  91499   
3  Line 1: 78266 Lien Road Line 2: PO Box 68671   PA-Reading  19610   
4      Line 1: 4392 Duke Place Line 2: Suite 79   WA-Spokane  99215   

                  full_name last_name, first_name middle_name         dob  \
0  Sanchis, Jerrylee Terrel          Sanchis, Jerrylee Terrel  11/20/1924   
1  Shapera, Florenza Jethro          Shapera, Florenza Jethro   3/14/1923   
2        Mart, Agustin Juan                Mart, Agustin Juan  12/19/1941   
3       Ticic, Pammi Renell               Ticic, Pammi Renell    8/2/1979   
4      Sprasen, Conn Tybalt              Sprasen, Conn Tybalt   4/14/1966   

                         email                                  guid  \
0          tsanchis0@google.cn  38b23627-9118-4180-b0d5-1396b8c99056   
1  jshapera1@howstuffworks.com  70f362c6-ad87-4e31-aa51-1f829fcdcebb   
2           jmart2@yahoo.co.jp  6b4cf83d-a760-4546-9b2a-2bbe146007f3   
3              rticic3@php.net  31c21b5d-7cdd-4acd-8592-b156b2e519c7   
4    tsprasen4@dagondesign.com  ecd3fec5-e4ec-45ff-91b0-bb0db86596a9   

                        line1         line2      city state  
0        4 Continental Trail                   Peoria    IL  
1  0259 Messerschmidt Street                     Erie    PA  
2        3461 Bartillon Lane       Suite 26  Van Nuys    CA  
3            78266 Lien Road   PO Box 68671   Reading    PA  
4            4392 Duke Place       Suite 79   Spokane    WA

In [127]:
# Convert the date of birth column to a datetime object:i
df3['dob'] = pd.to_datetime(df3['dob'])
df3.head(5)

id first_name middle_name last_name  \
0  1.0   Jerrylee      Terrel   Sanchis   
1  2.0   Florenza      Jethro   Shapera   
2  3.0    Agustin        Juan      Mart   
3  4.0      Pammi      Renell     Ticic   
4  5.0       Conn      Tybalt   Sprasen   

               Line 1: address Line 2: address2   state-city    zip  \
0          Line 1: 4 Continental Trail Line 2:     IL-Peoria  61614   
1    Line 1: 0259 Messerschmidt Street Line 2:       PA-Erie  16565   
2  Line 1: 3461 Bartillon Lane Line 2: Suite 26  CA-Van Nuys  91499   
3  Line 1: 78266 Lien Road Line 2: PO Box 68671   PA-Reading  19610   
4      Line 1: 4392 Duke Place Line 2: Suite 79   WA-Spokane  99215   

                  full_name last_name, first_name middle_name        dob  \
0  Sanchis, Jerrylee Terrel          Sanchis, Jerrylee Terrel 1924-11-20   
1  Shapera, Florenza Jethro          Shapera, Florenza Jethro 1923-03-14   
2        Mart, Agustin Juan                Mart, Agustin Juan 1941-12-19   
3       Ticic, Pammi Renell               Ticic, Pammi Renell 1979-08-02   
4      Sprasen, Conn Tybalt              Sprasen, Conn Tybalt 1966-04-14   

                         email                                  guid  \
0          tsanchis0@google.cn  38b23627-9118-4180-b0d5-1396b8c99056   
1  jshapera1@howstuffworks.com  70f362c6-ad87-4e31-aa51-1f829fcdcebb   
2           jmart2@yahoo.co.jp  6b4cf83d-a760-4546-9b2a-2bbe146007f3   
3              rticic3@php.net  31c21b5d-7cdd-4acd-8592-b156b2e519c7   
4    tsprasen4@dagondesign.com  ecd3fec5-e4ec-45ff-91b0-bb0db86596a9   

                        line1         line2      city state  
0        4 Continental Trail                   Peoria    IL  
1  0259 Messerschmidt Street                     Erie    PA  
2        3461 Bartillon Lane       Suite 26  Van Nuys    CA  
3            78266 Lien Road   PO Box 68671   Reading    PA  
4            4392 Duke Place       Suite 79   Spokane    WA

In [128]:
# Validate the data types for each column:
df3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1000 entries, 0 to 999
Data columns (total 16 columns):
 #   Column                             Non-Null Count  Dtype         
---  ------                             --------------  -----         
 0   id                                 1000 non-null   float64       
 1   first_name                         1000 non-null   object        
 2   middle_name                        1000 non-null   object        
 3   last_name                          1000 non-null   object        
 4   Line 1: address Line 2: address2   1000 non-null   object        
 5   state-city                         1000 non-null   object        
 6   zip                                1000 non-null   int64         
 7   full_name                          1000 non-null   object        
 8   last_name, first_name middle_name  1000 non-null   object        
 9   dob                                1000 non-null   datetime64[ns]
 10  email                              10

In [129]:
# Convert the date specified by Eli to a datetime object:
from datetime import datetime
date_string = "3/21/2023"
date_format = "%m/%d/%Y"
date_obj = datetime.strptime(date_string, date_format)
print(date_obj)

2023-03-21 00:00:00


In [130]:
# Add an age column by subtracting our date from the date of birth:
# 
df3['age'] = (date_obj - df3['dob']).astype('<m8[Y]')
df3.head(5)

id first_name middle_name last_name  \
0  1.0   Jerrylee      Terrel   Sanchis   
1  2.0   Florenza      Jethro   Shapera   
2  3.0    Agustin        Juan      Mart   
3  4.0      Pammi      Renell     Ticic   
4  5.0       Conn      Tybalt   Sprasen   

               Line 1: address Line 2: address2   state-city    zip  \
0          Line 1: 4 Continental Trail Line 2:     IL-Peoria  61614   
1    Line 1: 0259 Messerschmidt Street Line 2:       PA-Erie  16565   
2  Line 1: 3461 Bartillon Lane Line 2: Suite 26  CA-Van Nuys  91499   
3  Line 1: 78266 Lien Road Line 2: PO Box 68671   PA-Reading  19610   
4      Line 1: 4392 Duke Place Line 2: Suite 79   WA-Spokane  99215   

                  full_name last_name, first_name middle_name        dob  \
0  Sanchis, Jerrylee Terrel          Sanchis, Jerrylee Terrel 1924-11-20   
1  Shapera, Florenza Jethro          Shapera, Florenza Jethro 1923-03-14   
2        Mart, Agustin Juan                Mart, Agustin Juan 1941-12-19   
3       Ticic, Pammi Renell               Ticic, Pammi Renell 1979-08-02   
4      Sprasen, Conn Tybalt              Sprasen, Conn Tybalt 1966-04-14   

                         email                                  guid  \
0          tsanchis0@google.cn  38b23627-9118-4180-b0d5-1396b8c99056   
1  jshapera1@howstuffworks.com  70f362c6-ad87-4e31-aa51-1f829fcdcebb   
2           jmart2@yahoo.co.jp  6b4cf83d-a760-4546-9b2a-2bbe146007f3   
3              rticic3@php.net  31c21b5d-7cdd-4acd-8592-b156b2e519c7   
4    tsprasen4@dagondesign.com  ecd3fec5-e4ec-45ff-91b0-bb0db86596a9   

                        line1         line2      city state    age  
0        4 Continental Trail                   Peoria    IL   98.0  
1  0259 Messerschmidt Street                     Erie    PA  100.0  
2        3461 Bartillon Lane       Suite 26  Van Nuys    CA   81.0  
3            78266 Lien Road   PO Box 68671   Reading    PA   43.0  
4            4392 Duke Place       Suite 79   Spokane    WA   56.0

In [131]:
# Clean up the dataframe by dropping unneeded columns:
# 
# There are two ways to drop columns: 
# df3.drop(['last_name,  firstname middle_name'], axis=1)
# df3.drop(columns=['last_name, first_name middle_name'])
df4 = df3.drop(columns=['Line 1: address Line 2: address2', 'state-city', 'full_name', 'last_name, first_name middle_name'])
df4.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1000 entries, 0 to 999
Data columns (total 13 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   id           1000 non-null   float64       
 1   first_name   1000 non-null   object        
 2   middle_name  1000 non-null   object        
 3   last_name    1000 non-null   object        
 4   zip          1000 non-null   int64         
 5   dob          1000 non-null   datetime64[ns]
 6   email        1000 non-null   object        
 7   guid         1000 non-null   object        
 8   line1        1000 non-null   object        
 9   line2        1000 non-null   object        
 10  city         1000 non-null   object        
 11  state        1000 non-null   object        
 12  age          1000 non-null   float64       
dtypes: datetime64[ns](1), float64(2), int64(1), object(9)
memory usage: 109.4+ KB


In [111]:
# Create new dataframe using ages between 40 and 60, inclusive:
age_df = df4.loc[(df4["age"] >= 40) & (df4["age"] <= 60) , :]
age_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 268 entries, 3 to 999
Data columns (total 13 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   id           268 non-null    float64       
 1   first_name   268 non-null    object        
 2   middle_name  268 non-null    object        
 3   last_name    268 non-null    object        
 4   zip          268 non-null    int64         
 5   dob          268 non-null    datetime64[ns]
 6   email        268 non-null    object        
 7   guid         268 non-null    object        
 8   line1        268 non-null    object        
 9   line2        268 non-null    object        
 10  state        268 non-null    object        
 11  city         268 non-null    object        
 12  age          268 non-null    float64       
dtypes: datetime64[ns](1), float64(2), int64(1), object(9)
memory usage: 29.3+ KB


Print mailing address for everyone who is of age 40 to 60 inclusive as of 3/21/2023

Pammi Rennel Ticic  
78266 Lien Road  
PO Box 68671  
Reading, PA 19610

Also, if there is no line 2: or line 2 is blank, there should not be a blank

Bob Smith  
866 Maple Ave.  
Chicago, IL 60062

In [112]:
# Print out the addresses:
for index, row in age_df.iterrows():
    print(f"{row['first_name']} {row['middle_name']} {row['last_name']} ")
    print(f"{row['line1']}")
    if row['line2'] != "": 
        print(f"{row['line2']}")
    print(f"{row['city']}, {row['state']} {row['zip']} ")
    print('----')

Pammi Renell Ticic 
78266 Lien Road 
PO Box 68671
Reading, PA 19610 
----
Conn Tybalt Sprasen 
4392 Duke Place 
Suite 79
Spokane, WA 99215 
----
Candide Delmore Tilliard 
037 Victoria Parkway 
Boston, MA 2203 
----
Yovonnda Edward Maddrah 
7 Nova Drive 
Newark, NJ 7112 
----
Tait Tull Gullis 
43 Maple Wood Junction 
Atlanta, GA 30301 
----
Aymer Hunfredo McCreery 
8 Artisan Drive 
Phoenix, AZ 85083 
----
Ariana Mair Comiam 
7 Hoard Circle 
PO Box 34663
Atlanta, GA 31119 
----
Mirabelle Davidson Bonney 
24 Sherman Place 
Birmingham, AL 35244 
----
Zola Livvie Pogson 
536 Petterle Park 
Los Angeles, CA 90065 
----
Parnell Isa Dicks 
17 Holmberg Way 
Tucson, AZ 85705 
----
Tammie Davie Foucar 
5864 Artisan Road 
7th Floor
Saint Augustine, FL 32092 
----
Ambros Victoria Whitesel 
62332 Kropf Avenue 
Apt 1299
Saint Louis, MO 63126 
----
Travers Myra Reckhouse 
53 Meadow Valley Alley 
PO Box 18012
San Diego, CA 92196 
----
Dorri Norene Gerant 
9 Fieldstone Street 
Apt 1604
Sioux Falls, SD 57